### PR TITLE
MAINT: remove incidental imports from private modules

### PR DESCRIPTION
**Pull request by h-vetinari**
This has now grown into a full-blown fix for #19905; while we're targetting 1.14 with this, I wanted to check if we're actually warning in all cases (after discovering some potential discrepancies).

___
_Previously:_

Some more follow-ups from the work for https://github.com/scipy/scipy/issues/18279, same as #19893.

I've tried & failed to import these on 1.11 resp 1.12; the imports are fully ignored AFAICT due to the presence of `__all__` in those private modules.

Not sure what the intention there was @j-bowhay, but in any case, this should be good to remove IMO.

### CODE REVIEW

**Comment by mdhaber** [comment]
These imports were intentional and part of the plan to fix gh-18279. The current behavior is:
```python3
from scipy.optimize.slsqp import exp
```

which produces the desired warning message.
```
DeprecationWarning: `scipy.optimize.slsqp.exp` is deprecated along with the `scipy.optimize.slsqp` namespace. `scipy.optimize.slsqp.exp` will be removed in SciPy 1.13.0, and the `scipy.optimize.slsqp` namespace will be removed in SciPy 2.0.0.
```

**Comment by h-vetinari** [comment]
> The current behavior is, e.g.:

In one of these [discussions](https://github.com/scipy/scipy/pull/18992#discussion_r1279034139) the reason given for keeping these imports was that they once were part of `__all__`. As it happens, that wasn't the case anymore for the import I removed in https://github.com/scipy/scipy/pull/19893, but still affects some of the functions here.

> These imports were intentional and part of the plan to fix [gh-18279](https://github.com/scipy/scipy/issues/18279).

In any case, we've scheduled them for removal in 1.13, so the timing is still OK. 🙃

**Comment by mdhaber** [comment]
So yeah, go ahead and remove the tests. It's fine with me to take all these out now, even if 1.13 is coming earlier than usual.

**Comment by h-vetinari** [comment]
> In any case, we've scheduled them for removal in 1.13, so the timing is still OK. 🙃

Actually, we've delayed this to 1.14 in #19892. So I'll wait until 1.13 is out.

File: scipy/integrate/vode.py
Comment (j-bowhay) [inline scipy/integrate/vode.py:5]: In the meantime, one thing to think about is these empty modules. On the one hand, we said 2.0 but if they are empty...

Comment (j-bowhay) [reply scipy/integrate/vode.py:5]: Actually, upon reflection probably worth keeping as it provides a more helpful warning

**Comment by h-vetinari** [comment]
Lint error is not due to this PR:
```
scipy/optimize/_minpack_py.py:875:89: E501 Line too long (89 > 88)
```
https://github.com/scipy/scipy/blob/65bb18c07795e887803fe65522ce26f4b9a0888f/scipy/optimize/_minpack_py.py#L877-L881

**Comment by h-vetinari** [comment]
This is ready for review now (even though my memory of why I did what I did is a bit hazy two months later 😅). We should also double-check the discussion from https://github.com/scipy/scipy/issues/19905 w.r.t. `scipy.optimize.slsqp.approx_jacobian` resp. `scipy.io.mmio.MMFile`.

**Comment by dschmitz89** [comment]
@h-vetinari  The last two test failures are related. After fixing, this can be merged IMO.

**Comment by j-bowhay** [comment]
> This is ready for review now (even though my memory of why I did what I did is a bit hazy two months later 😅). We should also double-check the discussion from #19905 w.r.t. `scipy.optimize.slsqp.approx_jacobian` resp. `scipy.io.mmio.MMFile`.

We could consider leaving them in for now? I don't quite have the capacity at the moment to investigate properly what to do with these

**Comment by j-bowhay** [comment]
There is also #19223 to consider

**Comment by h-vetinari** [comment]
I'm not touching `MMFile` here, and `approx_jacobian` got removed partly because you had mentioned in https://github.com/scipy/scipy/issues/19905 that it was an accidental export as well. I don't mind deferring if you think it's worth delaying.

**Comment by j-bowhay** [comment]
> I'm not touching `MMFile` here, and `approx_jacobian` got removed partly because you had mentioned in #19905 that it was an accidental export as well. I don't mind deferring if you think it's worth delaying.

Ok cool that's fine then, I `approx_jacobian` was fine to remove its just the `io` stuff I can't remember all the details on

**Comment by h-vetinari** [comment]
@j-bowhay this is mainly a follow-up to your efforts around removing those incidental imports - I think it's ready and would appreciate your review (I think we were close already a month ago; I suggest to finish this now).

**Comment by j-bowhay** [comment]
Both this and #19673 are very much still on my radar I just need to finish a few exams and assignments before I can spend more than 10 minutes looking at something! I will get to this in May which I assume is fine with the 1.14 timeline.

File: scipy/_lib/tests/test_public_api.py
Comment (j-bowhay) [inline scipy/_lib/tests/test_public_api.py:456]: Perhaps we should consider failing the test if we hit this branch so to ensure we haven't missed anything

Comment (j-bowhay) [reply scipy/_lib/tests/test_public_api.py:456]: ie
```
attr = getattr(correct_import, attr_name, None)
assert attr is not None
message = f"Please import `{attr_name}` from the `{import_name}`..."
        with pytest.deprecated_call(match=message):
```

Comment (h-vetinari) [reply scipy/_lib/tests/test_public_api.py:456]: Done. 👍

File: scipy/io/matlab/mio5.py
Comment (j-bowhay) [inline scipy/io/matlab/mio5.py:16]: Can we leave `'varmats_from_mat'` in and skip it in the test?  #19223 is unresolved so think it would be safer to sit on it and get everything else merged

Comment (h-vetinari) [reply scipy/io/matlab/mio5.py:16]: I looked at doing that, but I don't want to add special handling in `test_public_api.py`. We should be dogfooding here - I've put it in the `scipy.io.matlab` namespace (which solves the warning confusion portion of #19223). Given that this was an explicit recommendation in the warning, I think we cannot really consider the function to be really private anymore anyway...

If we don't like the name (I certainly think it's... suboptimal), we should come up with a better alternative before the next release.

Comment (j-bowhay) [reply scipy/io/matlab/mio5.py:16]: Ok that's fine with me too. Two final points then I think this is good to go, can you update the version here
https://github.com/scipy/scipy/blob/12a231cea3cfa3e46102da7bf62ac5e138c3e8d8/scipy/_lib/deprecation.py#L61
so that we can remove `varmats_from_mat` from `scipy.io.matlab.mio5` in the future and could you post to the development forum as we are now adding to the public api

Comment (h-vetinari) [reply scipy/io/matlab/mio5.py:16]: > as we are now adding to the public api

Since our warnings have been pointing to them, the function must be considered public already. I really dislike its name, but that's what it is.

We can immediately redeprecate it if we want, as long as we provide the functionality somewhere else. In any case, I don't think we'll fix this in time for 1.14, so I'm tending back to your suggestion to skip it in the test and unblock the rest.

Comment (j-bowhay) [reply scipy/io/matlab/mio5.py:16]: Either is fine with me (I am not familiar enough with the `io` submodule to put forward a better name!)

**Comment by j-bowhay** [comment]
It looks like you've accidentally picked up a submodule change

**Comment by h-vetinari** [comment]
While I generally ask not to, this PR can (or even: should) be squash-merged.

**Comment by h-vetinari** [comment]
The failure here is unrelated:
```
 =========================== short test summary info ===========================
FAILED scipy\optimize\tests\test_constraint_conversion.py::TestNewToOld::test_individual_constraint_objects
= 1 failed, 63629 passed, 3042 skipped, 246 xfailed, 14 xpassed in 648.81s (0:10:48) =
```

**Comment by h-vetinari** [comment]
Any further comments here @j-bowhay et al.? 🙃

**Review by j-bowhay — APPROVED** [review]
LGTM and thanks for tackling @h-vetinari

**Pull request merged**
